### DISCUSSION TITLE
Develop a method that determines the different valid combinations of workflow steps

### DISCUSSION THREAD

**Discussion by anujsinha3** [uw-ssec/MAWpy#99] [Ideas]
Currently, these are the different workflow steps for Mobility Analysis:

- Incremental Clustering
- Trace Segmentation Clustering
- Address Oscillation (Oscillation Collector)
- Update Stay Duration (Stay Duration Calculator)
- Combine Extracted Stays (Stay Integrator)

The task is to develop a method or provide an exhaustive list that determines the different valid combinations of workflow steps in a complete workflow.

Examples of valid workflows:
- Incremental Clustering --> Update Stay Duration
- Incremental Clustering --> Update Stay Duration --> Address Oscillation --> Update Stay Duration

**Comment by gracejia513** [comment]
I propose the following 5 workflows (adopted from an older paper done by Cynthia's previous student):

1. Incremental clustering --> stay duration calculator/update stay duration (purpose: check how not addressing oscillation affects the number of stays, trips, and trajectories inferred)
2. Incremental clustering --> stay duration calculator/update stay duration --> address oscillation --> stay duration calculator/update stay duration (purpose: see how addressing oscillation during post-processing affects the stays, trips, and trajectories inferred) 
3. Address oscillation --> incremental clustering --> stay duration calculator/update stay duration  (purpose: see how addressing oscillation during post-processing affects the stays, trips, and trajectories inferred)
4. trace segmentation clustering --> stay duration calculator/update stay duration (purpose: check how using trace segmentation clustering instead of incremental clustering affects the stays, trips, and trajectories inferred)
5. trace segmentation clustering --> incremental clustering --> stay duration calculator/update stay duration (purpose: check how trace segmentation together with incremental clustering helps to infer stays, trips, and trajectories)

If necessary, we can bring this up in todays' meeting for Cynthia's approval. 

Based on the discussion, this list is not comprehensive, but they could be default workflows built-in to the package so that if the user wants, the user can just run one line calling any of the workflows and no need to redefine the workflow. 

Constraints can happen if the user defined workflow violates the assumption of the ordering of the algorithms. For example, stay duration calculator must come AFTER clustering algorithm (duration can only be calculated after clusters have been identified). So if the user comes up with a workflow that have stay duration calculater before clustering, we can prompt an error message saying that the workflow is infeasible.
